core/sdk-core/src/main/java/software/amazon/awssdk/core/exception/ApiCallAttemptTimeoutException.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.exception;
+
+import software.amazon.awssdk.annotations.SdkPublicApi;
+
+/**
+ * Signals that an api call attempt could not complete within the specified timeout.
+ */
+@SdkPublicApi
+public class ApiCallAttemptTimeoutException extends SdkClientException {
```

Comment:
When will this be thrown to customers vs `ApiCallTimeoutException`? Will they be expected to handle both, if they configure both, or can we unify them in some way? Can we document how the timeout will be exposed in the setting itself?

Reply:
`ApiCallAttemptTimeoutException` will be thrown if all retries fail to complete within the configured `ApiCallAttemptTimeout` but it doesn't exceed the configured `ApiCallTimeout`.  We can probably unify them. Let me check what's the behavior in v1.

Reply:
Checked out v1 and looks like it's the same behavior as here. We are throwing [`HttpRequestTimeoutException`](https://github.com/aws/aws-sdk-java/blob/4734de6fb0f80fe5768a6587aad3b9d0eaec388f/aws-java-sdk-core/src/test/java/com/amazonaws/http/timers/request/UnresponsiveServerTests.java#L83) for api call attempt time out and `ClientExeuctionTImeoutException` for api call time out. Since they are thrown from different reasons, I'd say we keep them separate so that customers get a clear view of the root cause.